File: LeetCode/15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */

// @lc code=start
class Solution {
public:
#define ll long long
#define vi vector<int>
#define pii pair<int,int>
#define fr(i,n) for(int i=0;i<n;i++)
#define all(v) v.begin(),v.end()
#define mem(a,b) memset(a,b,sizeof(a))
#define MOD 1000000007
#define pb push_back
#define mp make_pair
#define pii pair<int,int>
#define vll vector<ll>
#define vc vector<char>
#define vs vector<string>

    vector<vector<int>> threeSum(vector<int>& nums) {

        int n=nums.size();

        vector<vector<int>> v;

        if(n<3)
        return v;

        sort(all(nums));

        for(int i=0;i<n-2;i++)
        {
            if(i>0 && nums[i]==nums[i-1])
            continue;

            int l=i+1;
            int r=n-1;

            while(l<r)
            {
                int sum=nums[i]+nums[l]+nums[r];

                if(sum==0)
                {
                    v.pb({nums[i],nums[l],nums[r]});
                    l++;
                    r--;

                    while(l<r && nums[l]==nums[l-1])
                    l++;

                    while(l<r && nums[r]==nums[r+1])
                    r--;
                }
                else if(sum>0)
                r--;
                else
                l++;
            }
        }

        return v;

        
    }



 

};
// @lc code=end

File: LeetCode/15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        # Sort the array
        nums.sort()

        # Initialize the list of triplets
        triplets = []

        # Loop through the array
        for i in range(len(nums)):

            # If the current element is greater than zero, we can stop
            if nums[i] > 0:
                break

            # If the current element is the same as the previous element, we can skip
            if i > 0 and nums[i] == nums[i-1]:
                continue

            # Set the left and right pointers
            l = i + 1
            r = len(nums) - 1

            # Loop through the array
            while l < r:

                # Compute the sum
                s = nums[i] + nums[l] + nums[r]

                # If the sum is less than zero, increment the left pointer
                if s < 0:
                    l += 1

                # If the sum is greater than zero, decrement the right pointer
                elif s > 0:
                    r -= 1

                # If the sum is zero, we have a triplet
                else:
                    triplets.append([nums[i], nums[l], nums[r]])

                    # Increment the left pointer
                    l += 1

                    # Decrement the right pointer
                    r -= 1

                    # Skip the duplicates
                    while l < r and nums[l] == nums[l-1]:
                        l += 1

                    # Skip the duplicates
                    while l < r and nums[r] == nums[r+1]:
                        r -= 1

        # Return the list of triplets
        return triplets
        
# @lc code=end

